File: linear_select.py
#!/usr/bin/python
#coding:utf8
import sort

def Swap(a, b):
  return b, a

'''
分治算法,数组中寻找第k小的数字
每组5个，分n/5组，每组进行排序，对n/5组的中位数中找到中位数
将数组分成<M =M >M 三部分 theta(n)
递归调用查找第k小的数字，根据不同部分的个数来选择在哪个部分进行遍历
'''
def LinearSelect(arr, start_idx, end_idx, k):
  n = end_idx - start_idx + 1
  '''
    算法时间复杂度接近o(1)
  '''
  if n < 20:
    sort.QuickSort(arr, start_idx, end_idx)
    return arr[k - 1]

  '''
  每组5个元素进行排序，排序时间复杂度相当于O(1)
  并将每组的中味数放在数组的前边，进行排序，然后选取中位数
  '''
  for i in range(0, n/5):
    sort.QuickSort(arr, i*5, i*5 + 4)
    arr[i], arr[i*5+2] = Swap(arr[i], arr[i*5+2])

  '''
  选取中位数，不需要这么写，递归的调用本方法即可
  sort.QuickSort(arr, 0, n/5)
  m = arr[2]
  '''
  m = LinearSelect(arr, 0, n/5, n/10)

  '''
  分割三部分
  '''
  small_n, equal_n = Partition(arr, m, 0, n - 1)
  if small_n > k:
    return LinearSelect(arr, 0, small_n - 1, k)
  elif small_n + equal_n < k:
    return LinearSelect(arr, small_n + equal_n, n - small_n - equal_n - 1, k - small_n - equal_n)
  else:
    return arr[k - 1];


'''
将数组分成三部分，<m =m >m，返回小于的个数和等于的个数
线性时间复杂度，并且不需要分割完毕再进行统计个数
'''
def Partition(arr, start_idx, end_idx):
  n = end_idx - start_idx
  i = start_idx
  j = end_idx

  while i < j:
    while i < j and arr[i] < m:
      i+=1

    '''
    如果arr[i]等于m，那么需要后移动，这里有点复杂，需要画一下
    '''
    if arr[i] <= m:
      k = i + 1
      while k <= j:
        if arr[k] < m:
          arr[i], arr[k] = Swap(arr[i], arr[k])
          i+=1
        elif arr[k] > m:
          arr[i], arr[k] = Swap(arr[i], arr[k])
          break;
        else:
          k += 1

    '''
    k超越j，说明已经分配完毕
    '''
    if k > j:
      return i,j-i+1

    while i < j and arr[j] > m:
      j-=1

    if i < j:
      arr[i], arr[j] = Swap(arr[i], arr[j])
      '''
      如果和m相同，那么需要后移
      '''
      if arr[i] < m:
        i+=1
      j-=1


if __name__ == '__main__':
  print('start...')
  arr = [156, 81, 127, 147, 43, 196, 46, 38, 181]
  k = len(arr)/2
  m = LinearSelect(arr, 0, len(arr) - 1, k)
  print m

  sort.QuickSort(arr, 0, len(arr) - 1)
  print arr[k -1]

File: linear_select.cpp

#include<random>
#include<iostream>
#include<vector>

#define LINEAR_SELECT_LIB 1

#define LENGTH(a) (sizeof(a)/sizeof(int))

extern void InsertSort(int arr[], int n);

void Partition(int arr[], int n, int m, int &, int &);


std::vector<int> RandomInteger(int n, int bound_low, int bound_high)
{
  std::default_random_engine generator;
  std::uniform_int_distribution<int> distribution(bound_low, bound_high);
  std::vector<int>arr;
  for(int i = 0; i < n; i++)
  {
    int number = distribution(generator);
    arr.push_back(number);
  }
  return arr;
}

void Swap(int &a, int &b)
{
  a = a ^ b;
  b = a ^ b;
  a = a ^ b;
}
/*
寻找第k个小的元素
*/
int LinearSelect(int arr[], int n, int k)
{
  int i;
  /*
  开销：theta(1)
  */
  if(n < 20)
  {
    InsertSort(arr, n);
    return arr[k-1];
  }

  /* 选取Median
  每组5个，使用插入排序
  开销：theta(n):每个排序最多比较10次，n/5个排序，开销是theta(n)级别
  */
  for(i = 0; i < n/5; i++)
  {
    InsertSort(arr + i * 5, 5);
    /*
    小技巧，将每组的中位数交换到最前面
    */
    Swap(arr[i], arr[i * 5 + 2]);
  }
  /*
  选取中位数
  */
  int m = LinearSelect(arr, n/5, n/10);
  /*
  将arr分成三部分，小于m，等于m，大于m
  */
  int small_n, equal_n;
  Partition(arr, n, m, small_n, equal_n);
  if(small_n > k)
  {
    return LinearSelect(arr, small_n, k);
  }
  else if(small_n + equal_n >= k)
  {
    return arr[k-1];
  }
  else
  {
    return LinearSelect(arr + small_n + equal_n, n - small_n - equal_n, k - equal_n - small_n);
  }
}

/*
一次遍历
*/
void Partition(int arr[], int n, int m, int &small_n, int &equal_n)
{
  int i, j;

  for(i = 0,j = n - 1; i < j;)
  {
    while(arr[i] < m && i < j)
    {
      i++;
    }

    while(arr[j] > m && i < j)
    {
      j--;
    }

    if(arr[i] == m)
    {
      int k;
      for(k = i+1; k <= j; k++)
      {
        if(arr[k] < arr[i])
        {
          Swap(arr[k], arr[i]);
          i++;
        }
        else if(arr[k] > arr[i])
        {
          Swap(arr[k], arr[i]);
          break;
        }
      }

      if(k > j)
      {
        small_n = i;
        equal_n = j-i+1;
        return;
        //std::cout<<"small:"<<i<<",equal:"<<j-i+1<<std::endl;
      }
    }

    if(i < j)
    {
      Swap(arr[i], arr[j]);
      if(arr[i] < m)
      {
        i++;
      }
      j--;
    }
  }

  /*
  计算small & equal
  */
  /*
  small_n = 0;
  equal_n = 0;
  for(i = 0; i < n; i++)
  {
    if(arr[i] < m)
    {
      small_n++;
    }

    else if(arr[i] == m)
    {
      equal_n++;
    }
    else
    {
      break;
    }
  }
  */
}

#if LINEAR_SELECT_LIB < 1
int main(int argc, char **argv)
{
  int arr[] = {0,1,2,3,4,5,6,7,8,8,9,10};

  int k = 9;
  int o = LinearSelect(arr, LENGTH(arr), k);
  std::cout<<"LinearSelect Kth is:"<<o<<std::endl;


  InsertSort(arr, LENGTH(arr));
  std::cout<<"real Kth is:"<<arr[k-1]<<std::endl;
  return 0;

  int small_n, equal_n;
  Partition(arr, LENGTH(arr), 0, small_n, equal_n);
  std::cout<<"small & equal:"<<small_n<<","<<equal_n<<std::endl;
  for(int i = 0; i < LENGTH(arr); i++)
  {
    std::cout<<arr[i]<<std::endl;
  }

  return 0;
}
#endif
